site24x7/oauth/oauth.go

```diff
@@ -0,0 +1,187 @@
+package oauth
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+)
+
+const (
+	tokenDomain = "https://accounts.zoho.com"
+	tokenApi    = "/oauth/v2/token"
+	timeout     = time.Second
+)
+
+func setRequestHeaders(request *http.Request, accessToken string) {
+	request.Header.Set("Content-Type", "application/json;charset=UTF-8")
+	request.Header.Set("Accept", "application/json; version=2.0")
+	request.Header.Set("Authorization", "Zoho-oauthtoken " + accessToken)
+}
+
+func getUrl(urlValues url.Values) string {
+	baseAccurl := fmt.Sprintf("%s%s", tokenDomain, tokenApi)
+	urlToReturn := fmt.Sprintf("%s?%s", baseAccurl, urlValues.Encode())
+	return urlToReturn
+}
+
+type Authenticator struct {
+	// Protects tkns.AccessToken (updated in refresh goroutine)
+	sync.Mutex
+	tkns      tokens
+	storePath string
+}
+
+func (ator *Authenticator) getClient() *http.Client {
+	trTls11 := &http.Transport{
+		DisableKeepAlives: true,
+		TLSClientConfig: &tls.Config{
+			MaxVersion:         tls.VersionTLS13,
+			InsecureSkipVerify: true,
+		},
+	}
+	client := http.Client{
+		Transport: trTls11,
+		Timeout:   timeout,
+	}
+	return &client
+}
+
+func (ator *Authenticator) setDefaultHeaders(request *http.Request) {
+	request.Header.Set("Accept", "application/x-www-form-urlencoded")
+}
+
+func (ator *Authenticator) getAccessTokenFrom(urlToPost string) error {
+	var result map[string]interface{}
+	req, err := http.NewRequest("POST", urlToPost, nil)
+	if err != nil {
+		return err
+	}
+	setRequestHeaders(req, ator.tkns.AccessToken)
+	client := ator.getClient()
+
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	err = json.NewDecoder(res.Body).Decode(&result)
+	if err != nil {
+		return err
+	}
+	ator.Lock()
+	ator.tkns.AccessToken = result["access_token"].(string)
```

Comment:
Is it guaranteed `result["access_token"]` always have a string type value at this point?

Reply:
Based on line 84, I don' think so?

Using this seems safer with type assertion:
```suggestion
	ator.tkns.AccessToken, _ = result["access_token"].(string)
```

Reply:
done